src/java/org/apache/cassandra/streaming/async/StreamingInboundHandler.java

```diff
@@ -0,0 +1,236 @@
+    private Thread blockingIOThread;
+
+    public StreamingInboundHandler(InetSocketAddress remoteAddress, int protocolVersion)
+    {
+        this.remoteAddress = remoteAddress;
+        this.protocolVersion = protocolVersion;
+        state = State.START;
+    }
+
+    @Override
+    @SuppressWarnings("resource")
+    public void handlerAdded(ChannelHandlerContext ctx)
+    {
+        bufChannel = new ByteBufReadableByteChannel(AUTO_READ_LOW_WATER_MARK, AUTO_READ_HIGH_WATER_MARK, ctx.channel().config());
+        blockingIOThread = new FastThreadLocalThread(new DeserializingSstableTask(ctx),
+                                                     String.format("Stream-Inbound--%s-%s", ctx.channel().id(), remoteAddress.toString()));
+        blockingIOThread.setDaemon(true);
+        blockingIOThread.start();
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object message) throws Exception
+    {
+        if (state == State.CLOSED)
+        {
+            ReferenceCountUtil.release(message);
+            return;
+        }
+
+        if (!(message instanceof ByteBuf))
+        {
+            ctx.fireChannelRead(message);
+            return;
+        }
+
+        bufChannel.append((ByteBuf) message);
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx)
+    {
+        close();
+        ctx.fireChannelInactive();
+    }
+
+    private void close()
+    {
+        if (state == State.CLOSED)
+            return;
+
+        state = State.CLOSED;
+        blockingIOThread.interrupt();
+        // TODO:JEB release resources;
+    }
+
+    @Override
+    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception
+    {
+        logger.error("exception occurred while in processing streaming file", cause);
+        close();
+        ctx.fireExceptionCaught(cause);
+    }
+
+    /**
+     * For testing
+     */
+    State getState()
+    {
+        return state;
+    }
+
+    /**
+     * For testing only!!
+     */
+    void setPendingBuffers(ByteBufReadableByteChannel bufChannel)
+    {
+        this.bufChannel = bufChannel;
+    }
+
+    /**
+     */
+    private class DeserializingSstableTask implements Runnable
+    {
+        private final ChannelHandlerContext ctx;
+
+        private StreamSession session;
+
+        DeserializingSstableTask(ChannelHandlerContext ctx)
+        {
+            this.ctx = ctx;
+        }
+
+        @Override
+        public void run()
+        {
+            try
+            {
+                while (state != State.CLOSED)
+                {
+                    StreamMessage message = StreamMessage.deserialize(bufChannel, protocolVersion, null);
+                    if (message == null)
+                        continue;
+
+                    // StreamInitMessage & IncomingFileMessage each start new channels, and IMF needs a session to be established a priori
+                    if (message instanceof StreamInitMessage)
```

Comment:
This is really a generic inbound message handler right? Are there potentially races here with multiple threads interacting with StreamResultFuture and StreamManager and the global state? Can they register different state and then end up with a different StreamResultFuture or session?

Ideally I think each stream session would get a dedicated control channel and a dedicated actor thread to orchestrate control messages. Completely decouple the whole thing and turn into a queue of events being processed serially. The only concurrency and parallelism is then files being transferred in parallel which is pretty clearly shared nothing other than resource limits which are generally pretty easy to reason about since there is no risk of deadlock.